File: Source/Utility/GC.cs
﻿using System.Collections.Generic;
using WCSharp.Api;

public static class GC
{
    public static List<Affix> GCAffixes = new List<Affix>();

    public static void RemoveTrigger(ref trigger t)
    {
        if (t == null) return;
        t.ClearActions();
        t.ClearConditions();
        t.Dispose();
        t = null;
    }

    public static void RemoveTimer(ref timer t)
    {
        if (t == null) return;
        t.Pause();
        t?.Dispose();
        t = null;
    }

    public static void RemoveEffect(ref effect e)
    {
        if (e == null) return;
        e?.Dispose();
        e = null;
    }

    public static void RemoveGroup(ref group g)
    {
        if (g == null) return;
        g.Clear();
        g?.Dispose();
        g = null;
    }

    public static void RemoveAffix(Affix affix)
    {
        if (affix == null) return;
        var index = GCAffixes.IndexOf(affix);
        if (index != -1) GCAffixes[index] = null;
        GCAffixes.Remove(affix);
    }

    public static void RemoveDictionary<K, V>(ref Dictionary<K, V> dict)
    {
        if (dict == null) return;
        dict.Clear();
        dict = null;
    }

    public static void RemoveList<T>(ref List<T> list)
    {
        if (list == null) return;
        list.Clear();
        list = null;
    }

    public static void RemoveFilterFunc(ref filterfunc filter)
    {
        if (filter == null) return;
        filter.Dispose();
        filter = null;
    }

    public static void RemoveTimerDialog(ref timerdialog td)
    {
        if (td == null) return;
        td.Dispose();
        td = null;
    }
}

File: src/Utility/GC.ts
import { Affix } from 'src/Affixes/Affix'
import { Effect, Group, Timer, Trigger } from 'w3ts'
import { safeArraySplice } from './ArrayUtils'

export class GC {
    public static GCAffixes: Affix[] = []

    public static RemoveTrigger = (t: Trigger | undefined) => {
        if (!t) return
        t.destroy()
    }

    public static RemoveTimer = (t: Timer | undefined) => {
        if (!t) return
        t.pause()
        t.destroy()
    }

    public static RemoveEffect = (e: Effect | undefined) => {
        if (!e) return
        e.destroy()
    }

    public static RemoveGroup = (g: Group | undefined) => {
        if (!g) return
        g.clear()
        g.destroy()
    }

    public static RemoveAffix = (affix: Affix | undefined) => {
        if (!affix) return
        const index = GC.GCAffixes.indexOf(affix)
        if (index !== -1) (GC.GCAffixes as any)[index] = null
        safeArraySplice(GC.GCAffixes, a => a === affix)
    }

    public static RemoveList<T>(list: T[] | undefined) {
        if (!list) return
        list.length = 0
    }

    public static RemoveFilterFunc = (filter: filterfunc | (() => boolean)) => {}

    public static RemoveTimerDialog = (td: timerdialog) => {
        if (!td) return
        DestroyTimerDialog(td)
    }
}
